### PR TITLE
Add toggleable code snippets

### PR DESCRIPTION
**Pull request by creuter**
* Uses Ace Editor for the text editor.
* Most of the snippet rendering code is reused from previous iterations:
  * https://github.com/thoughtbot/refills/commit/52a8ed402a9d717f660d5b7b366ac374614b24f3
  * https://github.com/thoughtbot/refills/commit/94866dc09e2f173972973acad4593b7e8f1f0bba

toggle

### CODE REVIEW

File: source/_code.html.erb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<div class="container">
+  <button aria-controls="example-snippets"
```
Comment (creuter) [inline source/_code.html.erb]: TODO before merging: These `example` ids should be pulled through from the `code_for` argument.

File: helpers/snippet_helpers.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    end
+
+    def language
+      :html
+    end
+  end
+
+  class ScssSnippet < Snippet
+    def path_segments
+      [SOURCE_DIR, "stylesheets", "refills", namespace, "_#{name.dasherize}.scss"]
```
Comment (creuter) [inline helpers/snippet_helpers.rb]: It's over by two characters, and wrapping would make this one harder to read.

File: source/_code.html.erb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<div class="container">
+  <button aria-controls="<%= name %>-snippets"
```
Comment (tysongach) [inline source/_code.html.erb]: I’m _so happy_ to see accessibility considerations being part of this rebuild.

Something to think about: Is the toggling of code a function of saving screen real estate and therefore the _viewing_ of that code? If so, do we need to the toggle the accessibility of the code to assistive technology (with ARIA here), rather than it just be accessible at all times?

If we were to go that route, then we’d simply need to hide (`aria-hidden`) the toggle button, since it would not need to be used with assistive technology.

This is one of those times where accessibility tools (like ARIA) and usability (with accessibility in mind) create a nuanced situation.

In its current state (hiding and showing the code examples), if we step back and take into consideration the rendering page/DOM as a whole, we have a page outline (the page's headings) of components with a title, but no indication the component's code is available. That is unless you parse all the way through the component and toggle the button.

I don’t think there is a right or wrong answer here, just something to think about.

File: source/_snippet.html.erb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+<div class="snippet">
+  <header class="snippet-header">
```
Comment (tysongach) [inline source/_snippet.html.erb]: I think we can remove the `header` here. If this is needed for styling, a `div` would be a solid replacement.

The reason is that `header` comes with an implicit role of `banner`, and every `header` element then gets surfaced in things like the rotor menu found in VoiceOver (think of it like a way to quickly navigate around a page's landmarks). It looks like this:

screen shot 2017-01-13 at 14 08 00

**Comment by tysongach** [comment]
I just noticed that with JavaScript disabled, the code snippets do not show. Should we consider enabling the toggling only if JavaScript is available, and if not, fallback to showing the code snippets?

**Comment by whmii** [comment]
Looks great, the only comment I would have is that I could see a toggled state to the button.

**Comment by creuter** [comment]
@tysongach I think overall the text editor is a much nicer reading experience.

It gives us some things for free, like like syntax highlighting and line numbers. But it also lets you click into it an use your arrow keys, highlights the current line, lets you `cmd+a` and use other selection shortcuts, etc.

File: source/_snippet.html.erb (hunk 4 above)
Comment (creuter) [inline source/_snippet.html.erb]: Just played with this, TIL.

I'll drop it altogether for now, and just keep the `h3`. We may want to reintroduce a wrapping `div` if we introduce a "copy code" button or something down the road.

File: source/_code.html.erb (hunk 3 above)
Comment (creuter) [inline source/_code.html.erb]: I see what you're getting at, played around with it some, and it's an interesting thought exercise.

I *think* (at least in this case) that a preferred experience for those using assistive technologies would emulate the experience of those who don't. That way there wouldn't be major confusing discrepancies with screen readers and what's visible on the page.

File: source/banner.html.erb [hunk 5]
```diff
@@ -1,4 +1,4 @@
-<section class="banner">
+<header class="banner" role-"banner">
```
Comment (tysongach) [inline source/banner.html.erb]:
```diff
-role-"banner"
+role="banner"
```

**Comment by creuter** [comment]
@tysongach @whmii 

I need to update the linter, but this is ready for further review.

* It's progressive, so the snippets are visible (and formatted) without js
* With js, the button is visible, and the text changes when toggling

screen shot 2017-01-20 at 10 57 19 am

toggle

File: source/_code.html.erb [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+          aria-expanded="false"
```
Comment (tysongach) [inline source/_code.html.erb]: In the spirit of progressive enhancement, should `aria-controls` and `aria-expanded` be added to the button via JavaScript, once it is available? In fact, should this button be injected via JavaScript?

Comment (creuter) [inline source/_code.html.erb]: It has a `display: none` which the js removes on setup. I believe that since screen readers ignore elements with a display of none, it's accomplishing the same thing?

Comment (tysongach) [inline source/_code.html.erb]: Correct. However, that is just the case with screen readers. It’s possible other assistive technologies treat that differently. And if we also treat CSS as a progressive enhancement, the button will be present whether it is functioning or not. It would be even more bullet-proof by starting with no button, and injecting one as technologies become available.

I acknowledge this is splitting hairs, but thought I’d mention! I’ll leave it up to you whether or not to take this approach.

Here is what this looks like with no CSS or JavaScript:

screen shot 2017-01-20 at 11 58 58

Comment (creuter) [inline source/_code.html.erb]: Makes sense! This is exactly the type of project to split hairs on, and I appreciate it. I'll give this a go

Comment (creuter) [inline source/_code.html.erb]: I spiked it out and got it working. 

I appreciate the idealism, but think treating the presence of *any* CSS as a progressive enhancement is unusual and probably not worth the tradeoffs for maintainability and readability.

**Comment by whmii** [comment]
A thumb for you sir.

…
👍

File: source/_snippet.html.erb [hunk 7]
```diff
@@ -0,0 +1,4 @@
+<div class="snippet">
+  <h3 class="snippet-title"><%= language %></h3>
```
Comment (tysongach) [inline source/_snippet.html.erb]: I’d suggest that these `h3` headings be removed, or somehow made more unique.

Right now, the HTML outline produced will include three `h3` elements _per component_. So once we have a page with more than one component, these headings become less descriptive and it is hard to parse which headings belong to which example.

Couple thoughts:
- Are individual headings needed?
- Could the `h3` be something like “Code for Example Component”?
- Could we utilize visually hidden text or `aria-label` to override how this is perceived in things like screen readers, so that it would be “HTML _for Example Component_” and not just “HTML”.
